cpp/src/arrow/array.cc

```diff
@@ -772,6 +773,105 @@ std::shared_ptr<Array> MakeArray(const std::shared_ptr<ArrayData>& data) {
   return out;
 }
 
+// ----------------------------------------------------------------------
+// Misc APIs
+
+namespace internal {
+
+std::vector<ArrayVector> RechunkArraysConsistently(
+    const std::vector<ArrayVector>& groups) {
+  if (groups.size() <= 1) {
+    return groups;
+  }
+  // Adjacent slices defining the desired rechunking
+  std::vector<std::pair<int64_t, int64_t>> slices;
+  // Total number of elements common to all array groups
+  int64_t total_length = -1;
+
+  {
+    // Compute a vector of slices such that each array spans
+    // one or more *entire* slices only
+    // e.g. if group #1 has bounds {0, 2, 4, 5, 10}
+    //     and group #2 has bounds {0, 5, 7, 10}
+    // then the computed slices are
+    //     {(0, 2), (2, 4), (4, 5), (5, 7), (7, 10)}
+    std::set<int64_t> bounds;
+    for (auto& group : groups) {
+      int64_t cur = 0;
+      bounds.insert(cur);
+      for (auto& array : group) {
+        cur += array->length();
+        bounds.insert(cur);
+      }
+      if (total_length == -1) {
+        total_length = cur;
+      } else {
+        // XXX Should we return an error code instead?
+        DCHECK_EQ(total_length, cur)
+            << "Array groups should have the same number of elements";
```

Comment:
Since this API is internal, it's not necessary. Reaching this code path would indicate an internal programming error by the Arrow developer. Should this code path ever be exposed in some way to user input, then returning an error code would make more sense